### PR TITLE
An initial, still experimental take on a ModelTextFieldEmbedder

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    def forward(self, text_field_input: Dict[str, torch.Tensor]) -> torch.Tensor:
+        arg_dict = {self._input_name: text_field_input}
+
+        # TODO(mattg): switch to using __call__ once pytorch supports dictionary return values.
+        return self._model.forward(**arg_dict)[self._output_name]
+
+    @classmethod
+    def from_params(cls, vocab: Vocabulary, params: Params) -> 'ModelTextFieldEmbedder':
+        # TODO(mattg): we probably need to load the saved model here somehow, and have some
```
Comment (DeNeutoy) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: You can do this using:
```
for param in model.parameters():
    param.requires_grad = False
```
And then adding a filter to the parameters we pass to the optimizer: `filter(lambda param: param.requires_grad, model.parameters())`. This is an extremely elegant aspect of Pytorch in my opinion.

File: allennlp/models/simple_tagger.py [hunk 2]
```diff
@@ -42,10 +42,10 @@ def __init__(self, vocab: Vocabulary,
         self.tag_projection_layer = TimeDistributed(Linear(self.stacked_encoder.get_output_dim(),
                                                            self.num_classes))
 
-    # pylint: disable=arguments-differ
     def forward(self,  # type: ignore
```
Comment (DeNeutoy) [inline allennlp/models/simple_tagger.py]: We give models which can be `TextFieldEmbedders` a `get_embedding_representations()` method, which calls forward and contains the unpacking logic for what is needed for the representation.

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        self._model = model
+        self._input_name = model_input_name
+        self._output_name = model_output_name
+        self._output_dim = model_output_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self._output_dim
+
+    def forward(self, text_field_input: Dict[str, torch.Tensor]) -> torch.Tensor:
```
Comment (DeNeutoy) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: This is overly restrictive. In order to get embeddings out of the SRL model, we also need to pass in a `IndexField`. I think this can just be `kwargs` or something which gets passed down to the Model?

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    model_output_name : ``str``
+        The key that we will look for in the model's output dictionary.  The model must have
+        included a token-level representation in its output dictionary for it to be useful here.
+    model_output_dim : ``int``
+        ``TextFieldEmbedders`` need to implement ``get_output_dim()``.  We could either make the
+        model implement some method that tells us what the output dim is, or we can take it as a
+        parameter.  We're opting for the second option here.
+    """
+    def __init__(self,
+                 model: Model,
```
Comment (DeNeutoy) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: Should we take a constructor arg here which specifies whether the `Model` parameters are trainable?

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    ----------
+    model : ``Model``
+        The model that we will use to embed the text field.
+    model_input_name : ``str``
+        The name the model assigns to the text field input we want to use.  For example, the
+        :class:`~allennlp.models.SimpleTagger` model takes a text input field named ``tokens`` -
+        that's the name that should be passed here.
+    model_output_name : ``str``
+        The key that we will look for in the model's output dictionary.  The model must have
+        included a token-level representation in its output dictionary for it to be useful here.
```
Comment (DeNeutoy) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: Not sure this is exactly the case - you could imagine using one of these as a full sentence encoder or something.

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    dictionary to return.
+
+    NOTE: This is still largely experimental and needs some real-world testing before I'm confident
+    this API is correct and actually works well.
+
+    Parameters
+    ----------
+    model : ``Model``
+        The model that we will use to embed the text field.
+    model_input_name : ``str``
```
Comment (DeNeutoy) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: This won't work in the general case, I think (see my note about the SRL model below) - in order to run the model forward, it must have everything required apart from the `labels` etc. Instead, we should just have `forward` take `kwargs` and unpack them into `forward` of the wrapped model.

Comment (matt-gardner) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: But then this doesn't match the `TextFieldEmbedder` API.  I think for those uses, the model should just take an `SRL` model as input, instead of a `TextFieldEmbedder`.

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: Not if it's a `TextFieldEmbedder`.

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: Great.  I think there should just be a function (in `nn.util`?) that's `make_untrainable(Module)`, or something, and that's where those lines live.  I can imagine a few scenarios where you would want to do something like this, and don't want to duplicate the code.

File: allennlp/models/simple_tagger.py (hunk 2 above)
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: The tradeoff there is that then the logic has to be duplicated in every model that does this.  The idea with this class is essentially to _be_ that `get_embedding_representations()` method, with minimal requirements on the model it's wrapping.

File: allennlp/modules/text_field_embedders/model_text_field_embedder.py (hunk 4 above)
Comment (DeNeutoy) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: I vote here in order to allow someone to construct this easily from a trained model they have in memory, rather than a saved model.

Comment (matt-gardner) [inline allennlp/modules/text_field_embedders/model_text_field_embedder.py]: Agreed, I was also leaning toward having it here.

**Comment by matt-gardner** [comment]
The motivation for this PR was ELMo, and ELMo is already in using a different mechanism.  Not sure there's any point for this anymore, so I'm closing this PR.
